File: python_sdk/src/reality_capture/specifications/training.py
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum


class TrainingS3DInputs(BaseModel):
    segmentations_3d: list[str] = Field(
        description="List of 3D models to train on.",
        alias="segmentations3D"
    )
    preset: Optional[str] = Field(default=None, description="Path to a preset")
    detector_name: str = Field(description="Name of the detector to train", alias="detectorName")


class TrainingS3DOutputs(BaseModel):
    detector: str = Field(description="Full detector information (name/version)")


class Segmentation3DTrainingModel(Enum):
    SPLATNET = "SPLATNet"


class PointCloudFeature(Enum):
    RGB = "RGB"
    NORMAL = "NORMAL"
    INTENSITY = "INTENSITY"


class TrainingS3DOptions(BaseModel):
    epochs: Optional[int] = Field(
        None, description="Number of time to iterate over the entire dataset", ge=1, le=100
    )
    spacing: Optional[float] = Field(
        None,
        description="Spacing of the pointcloud seen by the detector (in meters).",
        gt=0
    )
    model: Optional[Segmentation3DTrainingModel] = Field(None, description="Training Model architecture to use.")
    features: Optional[list[PointCloudFeature]] = Field(None, description="Features to use for the training.")
    version_number: Optional[str] = Field(
        None, 
        description="String representing the version number for the newly trained detector.",
        alias="versionNumber",
        pattern=r"^\d+(?:\.\d+)?$"
    )


class TrainingS3DOutputsCreate(Enum):
    DETECTOR = "detector"


class TrainingS3DSpecificationsCreate(BaseModel):
    inputs: TrainingS3DInputs = Field(description="Inputs")
    outputs: list[TrainingS3DOutputsCreate] = Field(description="Outputs")
    options: Optional[TrainingS3DOptions] = Field(None, description="Options")


class TrainingS3DSpecifications(BaseModel):
    inputs: TrainingS3DInputs = Field(description="Inputs")
    outputs: TrainingS3DOutputs = Field(description="Outputs")
    options: Optional[TrainingS3DOptions] = Field(None, description="Options")


class Segmentation3DPair(BaseModel):
    segmentation_3d_a: str = Field(
        alias="segmentation3DA",
        description="Reality data id of ContextScene pointing to a segmented 3D model (time A)"
    )
    segmentation_3d_b: str = Field(
        alias="segmentation3DB",
        description="Reality data id of ContextScene pointing to a segmented 3D model (time B)"
    )


class TrainingCD3DInputs(BaseModel):
    segmentation_3d_pairs: list[Segmentation3DPair] = Field(
        alias="segmentation3DPairs",
        description="List of paired segmented 3D scenes for change detection training."
    )
    preset: Optional[str] = Field(default=None, description="Path to a preset")
    detector_name: str = Field(description="Name of the detector to train", alias="detectorName")


class TrainingCD3DOutputs(BaseModel):
    detector: str = Field(description="Full detector information (name/version)")


class TrainingCD3DOutputsCreate(Enum):
    DETECTOR = "detector"


class TrainingCD3DOptions(BaseModel):
    epochs: Optional[int] = Field(
        None, description="Number of times to iterate over the entire dataset", ge=1, le=100
    )
    spacing: Optional[float] = Field(
        None,
        description="Spacing of the point cloud seen by the detector (in meters).",
        gt=0
    )
    features: Optional[list[PointCloudFeature]] = Field(None, description="Features to use for the training.")
    ignore_class: Optional[int] = Field(
        None, alias="ignoreClass",
        description="Class index to ignore during training."
    )
    version_number: Optional[str] = Field(
        None,
        description="String representing the version number for the newly trained detector.",
        alias="versionNumber",
        pattern=r"^\d+(?:\.\d+)?$"
    )


class TrainingCD3DSpecificationsCreate(BaseModel):
    inputs: TrainingCD3DInputs = Field(description="Inputs")
    outputs: list[TrainingCD3DOutputsCreate] = Field(description="Outputs")
    options: Optional[TrainingCD3DOptions] = Field(None, description="Options")


class TrainingCD3DSpecifications(BaseModel):
    inputs: TrainingCD3DInputs = Field(description="Inputs")
    outputs: TrainingCD3DOutputs = Field(description="Outputs")
    options: Optional[TrainingCD3DOptions] = Field(None, description="Options")

File: typescript/packages/reality-capture/src/specifications/training.ts
import { z } from "zod";

export const TrainingS3DInputsSchema = z.object({
  segmentations3D: z.array(z.string()).describe("List of 3D models to train on."),
  preset: z.string().optional().describe("Path to a preset"),
  detectorName: z.string().describe("Name of the detector to train"),
});
export type TrainingS3DInputs = z.infer<typeof TrainingS3DInputsSchema>;

export const TrainingS3DOutputsSchema = z.object({
  detector: z.string().describe("Full detector information (name/version)"),
});
export type TrainingS3DOutputs = z.infer<typeof TrainingS3DOutputsSchema>;

export enum Segmentation3DTrainingModel {
  SPLATNET = "SPLATNet"
}

export enum PointCloudFeature {
  RGB = "RGB",
  NORMAL = "NORMAL",
  INTENSITY = "INTENSITY"
}

export const TrainingS3DOptionsSchema = z.object({
  epochs: z
    .number().int()
    .gte(1, { message: "Must be greater than or equal to 1" })
    .lte(100, { message: "Must be less than or equal to 100" })
    .describe("Number of time to iterate over the entire dataset")
    .optional(),
  spacing: z
    .number()
    .gt(0, { message: "Must be greater than 0" })
    .describe("Spacing of the pointcloud seen by the detector (in meters).")
    .optional(),
  model: z
    .nativeEnum(Segmentation3DTrainingModel)
    .describe("Training Model architecture to use.")
    .optional(),
  features: z
    .array(z.nativeEnum(PointCloudFeature))
    .describe("Features to use for the training.")
    .optional(),
  versionNumber: z
    .string()
    .regex(/\d+(.\d+)?/)
    .describe("String representing the version number for the newly trained detector.")
    .optional(),
});
export type TrainingS3DOptions = z.infer<typeof TrainingS3DOptionsSchema>;

export enum TrainingS3DOutputsCreate {
  DETECTOR = "detector"
}

export const TrainingS3DSpecificationsCreateSchema = z.object({
  inputs: TrainingS3DInputsSchema.describe("Inputs"),
  outputs: z
    .array(z.nativeEnum(TrainingS3DOutputsCreate))
    .describe("Outputs"),
  options: TrainingS3DOptionsSchema.describe("Options").optional(),
});
export type TrainingS3DSpecificationsCreate = z.infer<typeof TrainingS3DSpecificationsCreateSchema>;

export const TrainingS3DSpecificationsSchema = z.object({
  inputs: TrainingS3DInputsSchema.describe("Inputs"),
  outputs: TrainingS3DOutputsSchema.describe("Outputs"),
  options: TrainingS3DOptionsSchema.describe("Options").optional(),
});
export type TrainingS3DSpecifications = z.infer<typeof TrainingS3DSpecificationsSchema>;